File: services/user_service.py
from models.user import User
from extensions import db
from werkzeug.security import generate_password_hash, check_password_hash

def register_user(username, password, email=None, role='user'):
    if User.query.filter_by(username=username).first():
        return None, '用户名已存在'
    if email and User.query.filter_by(email=email).first():
        return None, '邮箱已存在'
    password_hash = generate_password_hash(password)
    user = User(username=username, password_hash=password_hash, email=email, role=role)
    db.session.add(user)
    db.session.commit()
    return user, None

def verify_user(username, password):
    user = User.query.filter_by(username=username).first()
    if user and check_password_hash(user.password_hash, password):
        return user
    return None

def get_user_by_id(user_id):
    return User.query.get(user_id)

def get_user_by_username(username):
    return User.query.filter_by(username=username).first() 

File: backend-go/services/user_service.go
package services

import (
	"backend-go/database"
	"backend-go/models"
	"errors"

	"golang.org/x/crypto/bcrypt"
)

func RegisterUser(username, password, email, role string) (*models.User, error) {
	var existingUser models.User
	if err := database.DB.Where("username = ?", username).First(&existingUser).Error; err == nil {
		return nil, errors.New("用户名已存在")
	}

	if email != "" {
		if err := database.DB.Where("email = ?", email).First(&existingUser).Error; err == nil {
			return nil, errors.New("邮箱已存在")
		}
	}

	passwordHash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	user := models.User{
		Username:     username,
		PasswordHash: string(passwordHash),
		Email:        email,
		Role:         role,
		IsActive:     true,
	}

	if err := database.DB.Create(&user).Error; err != nil {
		return nil, err
	}

	return &user, nil
}

func VerifyUser(username, password string) (*models.User, error) {
	var user models.User
	if err := database.DB.Where("username = ?", username).First(&user).Error; err != nil {
		return nil, errors.New("用户名或密码错误")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(password)); err != nil {
		return nil, errors.New("用户名或密码错误")
	}

	return &user, nil
}

func GetUserByID(userID uint) (*models.User, error) {
	var user models.User
	if err := database.DB.First(&user, userID).Error; err != nil {
		return nil, err
	}
	return &user, nil
} 